<a href="https://colab.research.google.com/github/AlienX77-cmd/Python-for-Business-and-Finance/blob/main/Part_4_Exercise_03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Coding Exercises (Part 3)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment.
- Write Code completely __on your own__.

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas: Intermediate Topics


### Exercise 3: Analyzing Columns / Pandas Series

In [1]:
#run the cell!
import pandas as pd

In [2]:
#run the cell!
listings = pd.read_csv("listings.csv")

In [3]:
#run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [4]:
#run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6745 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   ADR TSO     140 non-null    float64
 6   IPO_Year    3105 non-null   float64
 7   Sector      5309 non-null   object 
 8   Industry    5309 non-null   object 
dtypes: float64(4), object(5)
memory usage: 481.9+ KB


19. __Select__ the numerical column __"Last_Price"__, create a __copy__ and __save__ the column/Pandas Series in the variable __price__! __Fill in the gaps!__

In [5]:
price = listings.Last_Price.copy()

In [6]:
#run the cell!
price.head()

0    81.68
1    29.15
2    75.39
3     2.16
4    34.02
Name: Last_Price, dtype: float64

20. Get some __summary statistics__ on the Series __price__! What is the __price__ of the __most expensive stock__?

In [7]:
price.describe()

count    6745.000000
mean       37.372232
std        99.249040
min         0.003000
25%         8.310000
50%        20.573500
75%        37.370000
max      4299.990000
Name: Last_Price, dtype: float64

The price of the most expensive stock is ... 4299.99!

21. Get the __minimum Value__ in the Series __price__ by explicitly calling the ... method! The cheapest stock has a __Last_Price__ of...?

In [8]:
price.min()

0.003

The cheapest stock has a stock price of... 0.003!

22. Get the Frequency/__Counts__ of all unique __values__ in the Series __price__! What is the __most frequent__ price?

In [9]:
price.value_counts()

25.0000     9
9.7300      9
1.6100      8
2.5000      8
10.2600     7
           ..
0.6712      1
103.3600    1
25.8850     1
7.7700      1
101.3300    1
Name: Last_Price, Length: 4470, dtype: int64

The most frequent values are ... 25.0 and 9.73!

23. Get the __relative frequencies__ in the Series __price__! What is the __relative frequency__ of the __most frequent price__?

In [10]:
price.value_counts(normalize = True)

25.0000     0.001334
9.7300      0.001334
1.6100      0.001186
2.5000      0.001186
10.2600     0.001038
              ...   
0.6712      0.000148
103.3600    0.000148
25.8850     0.000148
7.7700      0.000148
101.3300    0.000148
Name: Last_Price, Length: 4470, dtype: float64

The relative frequency of 25.0 and 9.73 is ... 0.001334!

24. __Sort__ the Series __price__ from __low to high__! What is the __second lowest__ price?

In [12]:
price.sort_values(ascending = True)

1676    0.003
3375    0.004
6439    0.005
6650    0.007
5684    0.010
        ...  
6170      NaN
6403      NaN
6551      NaN
6717      NaN
6839      NaN
Name: Last_Price, Length: 6852, dtype: float64

The second lowest value is ... 0.0040!

25. __Sort__ the Series __price__ from __high to low__ and __save the changes__ by setting the __inplace__ parameter to __True__!  <br>
__Fill in the gaps!__

In [13]:
price.sort_values(ascending = False, inplace = True)

__Inspect__ the __first 5 elements__ of price! What is the __second highest__ price?

In [14]:
#run the cell!
price.head()

5505    4299.9900
4500    2910.4200
350     1847.3300
815     1806.0000
631     1316.6999
Name: Last_Price, dtype: float64

The second highest price is ... 2910.42!

26. __Sort__ the Series __price__ by the __Index__ (ascending) and __save__ the changes!

In [15]:
price.sort_index(ascending = True, inplace = True)

In [16]:
#run the cell!
price.head(10)

0    81.68
1    29.15
2    75.39
3     2.16
4    34.02
5    29.90
6     2.48
7    53.54
8    12.30
9    44.88
Name: Last_Price, dtype: float64

27. __Select__ the non-numerical column __"Sector"__, create a __copy__ and __save__ the column/Pandas Series in the variable __sector__!

In [17]:
sector = listings.Sector.copy()

Inspect! The first element/sector is ... ?

In [18]:
#run the cell!
sector.head()

0       Capital Goods
1    Basic Industries
2          Technology
3         Health Care
4      Transportation
Name: Sector, dtype: object

The first element/sector is ... Capital Goods !

28. Call the __describe()__ method on the non-numerical Series __"sector"__! What is the __most frequent__ value/sector?

In [19]:
sector.describe()

count        5309
unique         12
top       Finance
freq         1059
Name: Sector, dtype: object

The most frequent value/sector is ... Finance!

29. Get all __unique values__ in the Series __sector__!

In [20]:
sector.unique()

array(['Capital Goods', 'Basic Industries', 'Technology', 'Health Care',
       'Transportation', 'Finance', 'Consumer Services', nan,
       'Consumer Durables', 'Consumer Non-Durables', 'Miscellaneous',
       'Public Utilities', 'Energy'], dtype=object)

30. Last but not least, __count the frequencies__ in the Series __sector__! How often does the value/sector __Transportation__ appear?

In [21]:
sector.value_counts()

Finance                  1059
Health Care               898
Consumer Services         826
Technology                619
Capital Goods             389
Basic Industries          314
Energy                    292
Public Utilities          264
Consumer Non-Durables     236
Miscellaneous             152
Consumer Durables         137
Transportation            123
Name: Sector, dtype: int64

The value/sector Transportation appears ... 123 times!

# Well Done!

------------------------------------------------------------

# Hints (Spoiler!)

19. Last_Price and copy() method

20. describe() method

21. min() method

22. value_counts() method

23. normalize parameter

24. sort_values() method

25. ascending & inplace parameter

26. sort_index() method; inplace parameter

27. sector = listings.---.---

28. sector.---

29. unique() method

30. value_counts() method